optimize RNN on CPU

xiaomengy opened the pull request:
This PR fused the matmul ops for input sequence together which helped improve the performance for all RNN layers.
In our test for a speech model, previously the GRU layer takes about 40.16% of the total inference time (about 77ms). While after this PR, the GRU layer takes about 27.73% of the total inference time (about 43ms).

Summary: optimize RNN on CPU

Differential Revision: D16113360

VitalyFedyunin commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -324,55 +349,97 @@ struct LayerOutput {
+      const Tensor& inputs,
+      const hidden_type& input_hidden,
+      const cell_params& params) const override {
+    if (inputs.device().is_cpu()) {
```
What is wrong with the same approach in case of the GPU?

BIT-silence replied:
I will apply it in the next PR for GPU. Originally on GPU we do matmul instead of linear for input first, and fuse the bias part in at::_thnn_fused_lstm_cell function. So I think it will be better to make change on CPU only first then apply the GPU change in case the PR to be too large.

VitalyFedyunin reviewed (changes requested):
Side note: Lots of unrelated formatting changes, really hard to read.

BIT-silence commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -236,71 +236,96 @@ template<typename hidden_type_tmpl, typename cell_params_tmpl>
+      const cell_params& params,
+      bool pre_compute_input = false) const override {
     if (input.is_cuda()) {
+      DCHECK(!pre_compute_input);
```
Done.

BIT-silence commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -236,71 +236,96 @@ template<typename hidden_type_tmpl, typename cell_params_tmpl>
-
-    return new_gate + input_gate * (hidden - new_gate);
+    const auto chunked_igates = pre_compute_input
+        ? input.chunk(3, 1)
```
Thanks for the comment. I'm not sure if I correctly understand that. Here I used in_place add_ for chuncked_hgates, which comes from params.linear_hh(hidden).chunk(3, 1). The input part goes to chuncked_igates which is only used as the argument of add_. So I guess it should be safe?

VitalyFedyunin replied:
in case of pre_compute_input ( btw is it better to rename it to past tense?):
```
 // contains vector of views over input
 const auto chunked_igates = pre_compute_input ? input.chunk(3, 1) : params.linear_ih(input).chunk(3, 1);
 
// ...

 // changing input tesnsor inplace!
const auto new_gate = chunked_igates[2].add(chunked_hgates[2].mul_(reset_gate)).tanh_();
 
 ```

BIT-silence replied:
I am a little confused here.  This line is "chunked_igates[2].add" which is not a in_place add_, right? And tanh_ is applied on the result of add() which is another tensor in my understanding.

VitalyFedyunin replied:
My bad. LGTM.

BIT-silence replied:
Thanks for the detailed review.

BIT-silence commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -236,71 +236,96 @@ template<typename hidden_type_tmpl, typename cell_params_tmpl>
+    const auto& cx = std::get<1>(hidden);
 
     if (input.is_cuda()) {
+      DCHECK(!pre_compute_input);
```
DONE

BIT-silence commented:
This PR fused the matmul ops for input sequence together which helped improve the performance for all RNN layers.
In our test for a speech model, previously the GRU layer takes about 40.16% of the total inference time (about 77ms). While after this PR, the GRU layer takes about 27.73% of the total inference time (about 43ms).

xiaomengy committed:
optimize RNN on CPU (#22512)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/22512

optimize RNN on CPU

Differential Revision: D16113360

fbshipit-source-id: 32bcbe72c3749e4500d8223791b009db03a95d3d

VitalyFedyunin reviewed (changes requested):
Still contains inplace modification of the input tensor.
